### PR TITLE
fix: conda env inside script

### PR DESCRIPTION
**Pull request by Jahysama**
### Description

Small fix which enables to run conda envs inside `script:`. Because path to conda was not absolute, python executable also wasn't defined correctly. This fixes the issue described in [#1811](https://github.com/snakemake/snakemake/issues/1811).

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### CODE REVIEW

**Comment by dlaehnemann** [comment]
Thanks for digging into this!

Would you feel comfortable adding your minimal example from #1811 as a test case? This would ensure that (i) this problem is fixed and (ii) no future changes break this, again.

Such a test case would be a folder here:
https://github.com/snakemake/snakemake/tree/main/tests

And an entry here:
https://github.com/snakemake/snakemake/blob/main/tests/tests.py

Existing test cases are always a good starting point for creating a new one.

**Comment by Jahysama** [comment]
Thank you for response!
I will add new test case as soon as i can :)

**Comment by Jahysama** [comment]
Finished tests, @dlaehnemann could you please take a look at my PR again, thank you!

**Comment by dlaehnemann** [comment]
As I'm not a code owner, I cannot provide the final code review. But I'll point @johanneskoester here.

**Comment by dlaehnemann** [comment]
@Jahysama : It somehow doesn't find the dummy package on Windows. I have no clue about Windows and Python on Windows, and didn't see anything obvious, where it might go wrong. Could you check if you have any idea?

Also, I'm not a code owner, so I cannot provide the final code review. But I'll point @johanneskoester here.

**Comment by Jahysama** [comment]
@dlaehnemann I think it's the way pip is called on windows; instead of doing it with `pip install -e dummy_package`, it should be done like `python -m pip install -e dummy_package`. It happens because there is no `pip` by default in windows `PATH`.
Source: https://stackoverflow.com/questions/23832586/how-to-run-pip-on-windows. I already applied this fix.

**Comment by dlaehnemann** [comment]
Perfect, thanks for the quick fix. Fingers crossed that the test now also completes on windows.

File: snakemake/script.py [hunk 1]
```diff
@@ -611,9 +611,9 @@ def write_script(self, preamble, fd):
 
     def _is_python_env(self):
         if self.conda_env is not None and ON_WINDOWS:
-            prefix = self.conda_env
+            prefix = os.path.join(self.conda_base_path, "envs", self.conda_env)
         elif self.conda_env is not None:
-            prefix = os.path.join(self.conda_env, "bin")
+            prefix = os.path.join(self.conda_base_path, "envs", self.conda_env, "bin")
```
Comment (johanneskoester) [inline snakemake/script.py:616]: Conda envs are not always named. If they are not named (which is actually the recommended default), self.conda_env is already the correct path. Hence, you need to check here whether it is named or not. I wonder how this actually worked in the test cases apart from the prebuilt one. Maybe first try self.conda_env alone, and if that fails, try prefixing it with the conda base path?

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: snakemake/script.py (hunk 1 above)
Comment (Jahysama) [reply snakemake/script.py]: Sorry for such a long delay.  i have added `if` statements to detect if  `self.conda_env` is a correct path and to add `self.conda_base_path` if necessary.

**Comment by Jahysama** [comment]
@dlaehnemann Hello, David! Sorry for bothering you, but could you please run execution of remaining tests, thank you in advance!

**Comment by dlaehnemann** [comment]
Is it intended that the script `touch.py` is not included in the changes?

**Comment by Jahysama** [comment]
Sorry about it, just added it to my fork, forgot to add it to git

**Comment by dlaehnemann** [comment]
Ah, and then `input.txt` should probably not be included, as it should be created by that rule, right?

**Comment by Jahysama** [comment]
Yes, it is not included, because it will be created by the first Snakefile rule. Although `input.txt` is included only in `expected-results` folder.

**Comment by dlaehnemann** [comment]
Ah, I see. Thanks for the explanation.

I just merged the latest main branch status into this branch, because this is required for merging anyways. So if you add further changes to this branch, make sure to first `git pull` this merge commit.

**Comment by Jahysama** [comment]
@dlaehnemann Hello again, David, and a happy new 2023! I tried a different approach: installing a dummy package before the snakemake run. Could you please execute the tests again, fingers crossed. Thank you in advance!

**Comment by moritzschaefer** [comment]
@Jahysama @johanneskoester I see that after so much great efforts (just read through #1811 and the comments here), the PR has stalled. 

As I am also affected by this bug now, I would be super greatful if you could revisit and finalize this.

**Comment by Jahysama** [comment]
It seems like new version of `eido` package was released causing error below. [Args for `validate_project()` were changed](https://github.com/pepkit/eido/blob/266d78850617e9f65f8bf7651b6a88c80d97529a/eido/validation.py#L61), it doesn't seem to use `exclude_case` arg in [`validate_object`](https://github.com/pepkit/eido/blob/dea59d64c14eeb32b236c25d3fee90702aa2584f/eido/validation.py#L26C38-L26C50) in older version anyway, instead it got replaced with `sample_name_colname`, which is not directly required in `validate_project()`. I think it is safe to remove `exclude_case` arg from [`snakemake/workflow.py`](https://github.com/snakemake/snakemake/blob/f1b8342757fc5554eb7a9ed5d393c376e9089c3c/snakemake/workflow.py#L1332C57-L1332C69)
```
 Full Traceback (most recent call last):
  File "/home/runner/work/snakemake/snakemake/snakemake/__init__.py", line 654, in snakemake
    workflow.include(
  File "/home/runner/work/snakemake/snakemake/snakemake/workflow.py", line 1231, in include
    exec(compile(code, snakefile.get_path_or_uri(), "exec"), self.globals)
  File "/tmp/snakemake-f7zdwzdv/Snakefile", line 8, in <module>
    expand("{sample}.txt", sample=pep.sample_table["sample_name"])
  File "/home/runner/work/snakemake/snakemake/snakemake/workflow.py", line 1331, in pepschema
    eido.validate_project(
TypeError: validate_project() got an unexpected keyword argument 'exclude_case'
```

**Review by johanneskoester — APPROVED** [review]

**Comment by johanneskoester** [comment]
Thanks a lot!

**Pull request merged**